File: mobile_client/lib/voice/voice_waveform.dart
import 'dart:math' as math;

import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';

import 'voice_state.dart';

/// Flowing multi-line audio waveform (ref: the Siri-style ribbon). Several
/// overlaid sine bands of slightly different frequency/phase, their amplitude
/// driven by [amplitude] (mic RMS while listening, playback envelope while
/// speaking). Coloured from the active [state]'s palette so it matches the orb.
class VoiceWaveform extends StatefulWidget {
  const VoiceWaveform({
    super.key,
    required this.state,
    required this.amplitude,
    this.height = 120,
  });

  final VoiceState state;
  final ValueListenable<double> amplitude;
  final double height;

  @override
  State<VoiceWaveform> createState() => _VoiceWaveformState();
}

class _VoiceWaveformState extends State<VoiceWaveform>
    with SingleTickerProviderStateMixin {
  late final AnimationController _ctrl = AnimationController(
    vsync: this,
    duration: const Duration(seconds: 8),
  )..repeat();

  @override
  void dispose() {
    _ctrl.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return SizedBox(
      height: widget.height,
      width: double.infinity,
      child: AnimatedBuilder(
        animation: Listenable.merge([_ctrl, widget.amplitude]),
        builder: (context, _) => CustomPaint(
          painter: _WavePainter(
            t: _ctrl.value * 8,
            amp: widget.amplitude.value.clamp(0.0, 1.0).toDouble(),
            state: widget.state,
          ),
        ),
      ),
    );
  }
}

class _WavePainter extends CustomPainter {
  _WavePainter({required this.t, required this.amp, required this.state});

  final double t;
  final double amp;
  final VoiceState state;

  @override
  void paint(Canvas canvas, Size size) {
    final mid = size.height / 2;
    final pal = state.palette;
    // Three bands tinted across the iridescent range (inner, mid, rim).
    final bands = <(Color, double, double, double)>[
      (pal[1], 1.0, 1.0, 0.0),   // mid colour, base freq
      (pal[0], 0.7, 1.6, 1.3),   // inner colour, higher freq, phase shift
      (pal[3], 0.5, 0.7, 2.4),   // rim colour, lower freq
    ];

    // Idle/connecting/thinking: gentle self-motion. Listening/speaking: swell
    // with amplitude so it visibly reacts to the voice.
    final reactive = state == VoiceState.listening ||
        state == VoiceState.speaking;
    // Boost sensitivity: a gentle curve (sqrt) lifts quiet speech, and a higher
    // ceiling makes loud speech swing wide. Reactive states are far more lively.
    final env = reactive
        ? math.max(0.18, math.sqrt(amp) * 1.35).clamp(0.0, 1.4)
        : 0.16 + 0.05 * math.sin(t * 2.0);

    for (final (color, ampScale, freqScale, phase) in bands) {
      final paint = Paint()
        ..style = PaintingStyle.stroke
        ..strokeWidth = 2.0
        ..strokeCap = StrokeCap.round
        ..color = color.withValues(alpha: 0.85);
      final path = Path();
      // Fewer steps = cheaper repaint (was 64) — smoother on device.
      const steps = 44;
      for (var i = 0; i <= steps; i++) {
        final x = size.width * i / steps;
        final nx = i / steps;
        // Taper amplitude toward the edges so the ribbon pinches at the ends.
        final taper = math.sin(nx * math.pi);
        final wave = math.sin(nx * math.pi * 2 * (2.2 * freqScale) +
            t * 3.0 * freqScale + phase);
        final y = mid +
            wave * taper * size.height * 0.38 * env * ampScale;
        if (i == 0) {
          path.moveTo(x, y);
        } else {
          path.lineTo(x, y);
        }
      }
      canvas.drawPath(path, paint);
    }
  }

  @override
  bool shouldRepaint(covariant _WavePainter old) =>
      old.t != t || old.amp != amp || old.state != state;
}

File: ios_app/JarvisCopilot/Copilot/Voice/VoiceWaveform.swift
import Foundation

/// The level/geometry model behind the flowing multi-line audio waveform (the
/// Siri-style ribbon) from `voice/voice_waveform.dart`. Several overlaid sine
/// bands of slightly different frequency/phase, their amplitude driven by the
/// mic RMS while listening / the playback envelope while speaking.
///
/// Only the DATA lives here — the `Canvas` that strokes it is the second-wave
/// agent's view. Keeping the maths out of the painter makes the envelope curve
/// (the part users actually notice) unit-testable.
enum VoiceWaveformModel {

    /// Points per band. Fewer steps = cheaper repaint (was 64 in Flutter).
    static let steps = 44

    /// Fraction of the view height one band may swing at full envelope.
    static let heightFraction = 0.38

    struct Band: Equatable, Sendable {
        /// Index into `VoiceState.palette` — bands are tinted across the
        /// iridescent range so the ribbon matches the orb.
        let paletteIndex: Int
        let ampScale: Double
        let freqScale: Double
        let phase: Double
    }

    static let bands: [Band] = [
        Band(paletteIndex: 1, ampScale: 1.0, freqScale: 1.0, phase: 0.0), // mid colour, base freq
        Band(paletteIndex: 0, ampScale: 0.7, freqScale: 1.6, phase: 1.3), // inner, higher freq
        Band(paletteIndex: 3, ampScale: 0.5, freqScale: 0.7, phase: 2.4), // rim, lower freq
    ]

    /// Listening/speaking swell with amplitude so the ribbon visibly reacts to
    /// the voice; idle/connecting/thinking have gentle self-motion instead.
    static func isReactive(_ state: VoiceState) -> Bool {
        state == .listening || state == .speaking
    }

    /// Vertical scale for this frame. A gentle sqrt curve lifts quiet speech and
    /// a higher ceiling makes loud speech swing wide.
    static func envelope(state: VoiceState, amplitude: Double, t: Double) -> Double {
        guard isReactive(state) else { return 0.16 + 0.05 * sin(t * 2.0) }
        let amp = min(max(amplitude, 0), 1)
        return min(max(0.18, sqrt(amp) * 1.35), 1.4)
    }

    /// Vertical offset as a FRACTION OF THE VIEW HEIGHT at each of `steps + 1`
    /// evenly-spaced x positions. The view draws `y = midY + offset * height`.
    /// Amplitude tapers toward the edges so the ribbon pinches at the ends.
    static func offsets(band: Band, t: Double, env: Double) -> [Double] {
        (0...steps).map { i in
            let nx = Double(i) / Double(steps)
            let taper = sin(nx * .pi)
            let wave = sin(nx * .pi * 2 * (2.2 * band.freqScale)
                           + t * 3.0 * band.freqScale + band.phase)
            return wave * taper * heightFraction * env * band.ampScale
        }
    }
}
